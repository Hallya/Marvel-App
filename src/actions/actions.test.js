import * as act from "./actions";
import * as help from "../helpers/helpers";
import * as reducer from "../reducers/reducers";

import fetchByLetter from "../../__mocks__/API_Responses/fetchByLetter.json";
import fetchCharacter from "../../__mocks__/API_Responses/fetchCharacter.json";
import fetchCharacters from "../../__mocks__/API_Responses/fetchCharacters.json";
import somethingFetched from "../../__mocks__/API_Responses/fetchComic.json";
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk'
// import fetchMock from 'fetch-mock';

// CATEGORIES
const catCharacters = "characters";
const catComics = "comics";

// ID
const idCharacter = 1009149;
const idComic = 17486;

// URL REQUEST ID
const requestCharacters = "characters?limit=20&offset=0";

// CONFIGURATION OF REDUX STORE AND REDUX THUNK
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
//FIXED DATE
Date = jest.fn(() => 1482363367071);

describe("Simples actions", () => {

  it("GET_SEARCHBAR_FOCUS:: should return an action without data", () => {
    const expectedAction = {
      type: "GET_SEARCHBAR_FOCUS"
    }

    expect(act.getSearchbarFocus()).toEqual(expectedAction)
  })

  it("LEAVE_SEARCHBAR_FOCUS:: should return an action without data", () => {
    const expectedAction = {
      type: "LEAVE_SEARCHBAR_FOCUS"
    }

    expect(act.leaveSearchbarFocus()).toEqual(expectedAction)
  })

  it("SET_DISPLAY_FALSE:: should return an action without data", () => {
    const expectedAction = {
      type: "SET_DISPLAY_FALSE",
    }

    expect(act.setFalseOnDisplay()).toEqual(expectedAction)
  })

  it("RESET_RELATED_DATA:: should return an action with data", () => {
    const expectedAction = {
      type: "RESET_RELATED_DATA",
      data: {
        data: 0
      },
      id: idComic
    }

    expect(act.resetRelatedData(idComic, {data: 0})).toEqual(expectedAction)
  })

  it("SET_DATA:: should return an action with an id and data", () => {
    const expectedAction = {
      type: "SET_DATA",
      id: idComic,
      data: {
        data: 0
      }
    }

    expect(act.setData(idComic, {data: 0})).toEqual(expectedAction)
  })

  it("SET_RELATED_DATA:: should return an action with an id and data", () => {
    const expectedAction = {
      type: "SET_RELATED_DATA",
      id: idComic,
      data: 0
    }

    expect(act.setRelatedData(idComic, {
      data: {
        results: 0
      }
    })).toEqual(expectedAction)
  })

  it("SET_PROFIL:: should return an action with an id and data", () => {
    const expectedAction = {
      type: "SET_PROFIL",
      id: idComic,
      data: 0
    }

    expect(act.setProfil(idComic, 0)).toEqual(expectedAction)
  })

  it("SET_VISIBILITY_FILTER:: should return an action with filter", () => {
    const expectedAction = {
      type: "SET_VISIBILITY_FILTER",
      filter: "visible"
    }

    expect(act.setVisibilityFilter("visible")).toEqual(expectedAction)
  })

  it("SET_INFO:: should return an action with info", () => {
    const expectedAction = {
      type: "SET_INFO",
      value: "info"
    }

    expect(act.setInfo("info")).toEqual(expectedAction)
  })

  it("SET_CURRENT_PAGE:: should return an action with data, category, requestId and current date", () => {
    const expectedAction = {
      type: "SET_CURRENT_PAGE",
      data: 0,
      category: 'comics',
      requestId: 0,
      receivedAt: Date()
    }

    expect(act.setCurrentPage(0, "comics", 0)).toEqual(expectedAction)
  })

  it("RECEIVE_POSTS:: should return an action with data, category, requestId and current date", () => {
    const expectedAction = {
      type: "RECEIVE_POSTS",
      posts: 0,
      category: 'comics',
      requestId: 0,
      receivedAt: Date()
    }

    expect(act.receiveFetch("comics", {
      data: {
        results: 0
      }
    }, 0)).toEqual(expectedAction)
  })

  it("REQUEST_RELATED_DATA:: should return an action with an ID", () => {
    const expectedAction = {
      type: "REQUEST_RELATED_DATA",
      id: idComic,
    }

    expect(act.requestRelatedData(idComic)).toEqual(expectedAction)
  })

  it("REQUEST_POSTS:: should return an action with a requestId and a category ", () => {
    const expectedAction = {
      type: "REQUEST_POSTS",
      requestId: 0,
      category: 'comics'
    }

    expect(act.requestFetch('comics', 0)).toEqual(expectedAction)
  })

  it("ADD_MORE:: should return an action with a category and data ", () => {
    const expectedAction = {
      type: "ADD_MORE",
      data: 0,
      category: 'comics'
    }

    expect(act.addMore('comics', {
      data: {
        results: 0
      }
    })).toEqual(expectedAction)
  })
})


describe("Asyncs actions", () => {

  
  it('Function fetchContent:: fetch content from API and return an Object', () => {

    fetch.mockResponse(JSON.stringify(somethingFetched));

    act.fetchContent().then(response => {
      expect(response).toMatchSnapshot()
    })
  })
})


describe("Helpers & sync action", () => {
  
  it('Function createRequestId:: should get return a part of an URL generated by conditions defined', () => {

    const expectedUrl = "characters?limit=30&offset=0&";
    const condition = {
      type: "browsing",
      category: "characters",
      idKnown: false
    }

    expect(help.createRequestId(
      (0).toString(),
      condition
    ))
      .toEqual(expectedUrl);
  })
  
  it('Function doubleRequest:: should check if previous request match new one and return a boolean', () => {

    const lastRequest = "boo";
    const newRequest = "far";

    expect(help.doubleRequest(newRequest, lastRequest)).toEqual(false);
  })
  
  it('Function getCache:: should check id match with object\'s id in an array', () => {

    const array = [
      {
        id: 1,
        foo: "bar"
      },
      {
        id: 22,
        bar: "foo"
      }
    ];

    expect(help.getCache(22, array)).toEqual({id: 22, bar: "foo"});
  })
  
  it('Function getConditions:: should return an object defining conditions describing how createRequestId func should behave', () => {

    const condition = {
      type: "browsing",
      category: "characters",
      idKnown: false
    };

    expect(help.getConditions('ch')).toEqual(condition);
  })
})


describe("Reducers", () => {
  
  it('Reducer "posts" | action "REQUEST_POSTS":: should return a new state with last request id fetched, set isFetching to "true" and error to "null"', () => {
    
    const oldState = {
      lastFetchId: {
        characters: null,
        comics: null,
        relatedData: null,
      },
      cached: [],
      isFetching: false,
    };

    const newState = {
      lastFetchId: {
        characters: requestCharacters,
        comics: null,
        relatedData: null,
      },
      cached: [],
      isFetching: true,
      error: null
    };

    expect(reducer.posts(oldState, act.requestFetch(catCharacters, requestCharacters))).toEqual(newState);
  })
  
  it('Reducer "posts" | action "RECEIVE_POSTS":: should return a new state with isFetching to "false", add a new object to "items" with unique id, data, and released date', () => {
    
    const oldState = {
      lastFetchId: {
        characters: null,
        comics: null,
        relatedData: null,
      },
      cached: [],
      isFetching: false,
    };

    const newState = {
      lastFetchId: {
        characters: null,
        comics: null,
        relatedData: null,
      },
      cached: [{
        id: requestCharacters,
        data: [0, 1, 2],
        receivedAt: Date()
      }],
      isFetching: false,
    };

    const data = {
      data: {
        results: [0, 1, 2]
      }
    }

    expect(reducer.posts(oldState, act.receiveFetch(catCharacters, data, requestCharacters))).toEqual(newState);
  })
  
  it('Reducer "posts" | action "SET_CURRENT_PAGE":: should return a new state with new requestId in the right category', () => {
    
    const oldState = {
      lastFetchId: {
        characters: null,
        comics: null,
        relatedData: null,
      },
      cached: [],
      isFetching: false,
    };

    const newState = {
      lastFetchId: {
        characters: requestCharacters,
        comics: null,
        relatedData: null,
      },
      cached: [],
      isFetching: false,
    };

    expect(reducer.posts(oldState, act.setCurrentPage(requestCharacters, catCharacters))).toEqual(newState);
  })

  it('Reducer "posts" | action "ERROR_POSTS":: should return a new state with informations about category\'s last error\n\n', () => {
    
    const error = "!!!"
    
    const oldState = {
      characters: {
        lastFetchId: null,
        items: [],
        isFetching: true,
      },
      comics: {
        lastFetchId: null,
        items: [],
        isFetching: false,
      }
    };

    const newState = {
       characters: {
         lastFetchId: null,
         items: [],
         isFetching: false,
         error: error
       },
       comics: {
         lastFetchId: null,
         items: [],
         isFetching: false,
       }
     };

    expect(reducer.posts(oldState, act.newError(catCharacters, error))).toEqual(newState);
  })
  
  it('Reducer "actualPage" | action "SET_CURRENT_PAGE":: should return a new state with request id, the current category, data and released date, ', () => {
    
    const data = {data: [{},{}]};
    
    const oldState = {
      id: null,
      data: [],
      offset: 20,
      receivedAt: Date(),
      isFocused: false
    };

    const newState = {
      id: requestCharacters,
      data: [{},{}],
      category: catCharacters,
      offset: 0,
      receivedAt: Date(),
      isFocused: false
     };


    expect(reducer.actualPage(oldState, act.setCurrentPage(requestCharacters, catCharacters, data))).toEqual(newState);
  })

  it('Reducer "actualPage" | action "REQUEST_POSTS":: should return a new state with isFetching true and category actualised', () => {
    
    const oldState = {
      id: null,
      data: [],
      offset: 0,
      receivedAt: null,
      isFocused: false
    };

    const newState = {
      id: null,
      data: [],
      offset: 0,
      receivedAt: null,
      isFocused: false,
      isFetching: true
    };

    expect(reducer.actualPage(oldState, act.requestFetch(catCharacters, requestCharacters))).toEqual(newState);
  })

  it('Reducer "actualPage" | action "RECEIVE_POSTS":: should return a new state with isFetching false, category actualised, new id and data and released date', () => {
    
    const data = {data:{results:[{}]}}
    const oldState = {
      id: null,
      data: [],
      offset: 0,
      receivedAt: null,
      isFocused: false,
      isFetching: true
    };

    const newState = {
      id: requestCharacters,
      category: catCharacters,
      data: [{}],
      offset: 0,
      receivedAt: Date(),
      isFocused: false,
      isFetching: false
    };

    expect(reducer.actualPage(oldState, act.receiveFetch(catCharacters, data, requestCharacters))).toEqual(newState);
  })

  it('Reducer "actualPage" | action "ADD_MORE":: should return a new state with new result added to previous', () => {
    
    const data = {data: {results:[{id: 1},{id: 2},{id: 3},{id: 4},{id: 5},{id: 6}]}}

    const oldState = {
      id: null,
      data: [{id: 1},{id: 2},{id: 3}],
      isFetching: true,
      offset: 0,
      receivedAt: null,
      isFocused: false
    };

    const newState = {
      data: [{id: 1},{id: 2},{id: 3},{id: 4},{id: 5},{id: 6}],
      id: null,
      isFetching: false,
      offset: 30,
      receivedAt: null,
      isFocused: false
    };

    expect(reducer.actualPage(oldState, act.addMore(catCharacters, data))).toEqual(newState);
  })

  it('Reducer "actualPage" | action "GET_SEARCHBAR_FOCUS":: should return a new state with isFocused †rue', () => {

    const oldState = {
      id: null,
      data: [],
      offset: {
        characters: 20,
        comics: 20
      },
      receivedAt: null,
      isFocused: false
    };

    const newState = {
     id: null,
      data: [],
      offset: {
        characters: 20,
        comics: 20
      },
      receivedAt: null,
      isFocused: true
    };

    expect(reducer.actualPage(oldState, act.getSearchbarFocus())).toEqual(newState);
  })

  it('Reducer "actualPage" | action "LEAVE_SEARCHBAR_FOCUS":: should return a new state with isFocused false\n\n', () => {

    const oldState = {
      id: null,
      data: [],
      offset: {
        characters: 20,
        comics: 20
      },
      receivedAt: null,
      isFocused: true
    };

    const newState = {
     id: null,
      data: [],
      offset: {
        characters: 20,
        comics: 20
      },
      receivedAt: null,
      isFocused: false
    };

    expect(reducer.actualPage(oldState, act.leaveSearchbarFocus())).toEqual(newState);
  })
  
  it('Reducer "actualProfil" | action "SET_PROFIL":: should return a new state with new id and data', () => {

    const action = {
      type: act.SET_PROFIL,
      id: 1017100,
      data: "foo"
    },

    oldState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: [],
        previousRelatedData: []
      }
    },

    newState = {
      id: 1017100,
      data: "foo",
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: [],
        previousRelatedData: []
      }
    };
    
    expect(reducer.actualProfil(oldState, action)).toEqual(newState);
  })
  
  it('Reducer "actualProfil" | action "SET_INFO":: should return a new state with infos that should be displayed', () => {

    const action = {
      type: act.SET_INFO,
      value: "foo"
     },

    oldState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: [],
        previousRelatedData: []
      }
    },

    newState = {
      id: null,
      data: null,
      infosDisplayed: "foo",
      relatedData: {
        isFetching: false,
        actualRelatedData: [],
        previousRelatedData: []
      }
    };
    
    expect(reducer.actualProfil(oldState, action)).toEqual(newState);
  })
  
  it('Reducer "actualProfil" | action "REQUEST_RELATED_DATA":: should return a new state with isFetching set to true and lastFetchId actualised', () => {

    const action = {
      type: act.REQUEST_RELATED_DATA,
      id: "foo"
     },

    oldState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: [],
        previousRelatedData: []
      }
    },

    newState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: true,
        lastFetchId: "foo",
        actualRelatedData: [],
        previousRelatedData: []
      }
    };
    
    expect(reducer.actualProfil(oldState, action)).toEqual(newState);
  })
  
  it('Reducer "actualProfil" | action "SET_RELATED_DATA":: should return a new state with isFetching set to false and related data added to "actual" and "previous" keys', () => {

    const action = {
      type: act.SET_RELATED_DATA,
      data: "bar",
      id: "foo"
     },

    oldState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: true,
        actualRelatedData: [],
        previousRelatedData: []
      }
    },

    newState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: "bar",
        previousRelatedData: [{
          id: "foo",
          data: "bar"
        }]
      }
    };
    
    expect(reducer.actualProfil(oldState, action)).toEqual(newState);
  })
  
  it('Reducer "actualProfil" | action "RESET_RELATED_DATA":: should return a new state with isFetching set to false and reset data from "previous" to "actual" key\n\n', () => {

    const action = {
      type: act.RESET_RELATED_DATA,
      data: "bar"
     },

    oldState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: true,
        actualRelatedData: [],
        previousRelatedData: []
      }
    },

    newState = {
      id: null,
      data: null,
      infosDisplayed: null,
      relatedData: {
        isFetching: false,
        actualRelatedData: "bar",
        previousRelatedData: []
      }
    };
    
    expect(reducer.actualProfil(oldState, action)).toEqual(newState);
  })
 
  it('Reducer "relatedData" | action "SET_DATA":: should return a new state with new data and set displayed to true', () => {

    const action = {
      type: act.SET_DATA,
      data: {data:{results:["bar"]}},
     },

    oldState = {
      data: null,
      displayed: false
    },

    newState = {
      data: "bar",
      displayed: true
    };
    
    expect(reducer.relatedData(oldState, action)).toEqual(newState);
  })
 
  it('Reducer "relatedData" | action "SET_DISPLAY_FALSE":: should return a new state with "displayed" set on false\n\n', () => {

    const action = {
      type: act.SET_DISPLAY_FALSE,
     },

    oldState = {
      data: null,
      displayed: true
    },

    newState = {
      data: null,
      displayed: false
    };
    
    expect(reducer.relatedData(oldState, action)).toEqual(newState);
  })
 
  it('Reducer "visibilityFilter" | action "SHOW_DESCRIPTION":: should return a new state with "description" set on it\'s opposite current value', () => {

    const action = {
      type: "SET_VISIBILITY_FILTER",
      filter: act.VisibilityFilters.SHOW_DESCRIPTION
     },

    oldState = {
      images: false,
      description: false
    },

    newState = {
      images: false,
      description: true
    };
    
    expect(reducer.visibilityFilter(oldState, action)).toEqual(newState);
  })
 
  it('Reducer "visibilityFilter" | action "SHOW_IMAGES":: should return a new state with "images" set on it\'s opposite current value', () => {

    const action = {
      type: "SET_VISIBILITY_FILTER",
      filter: act.VisibilityFilters.SHOW_IMAGES
     },

    oldState = {
      images: false,
      description: false
    },

    newState = {
      images: true,
      description: false
    };
    
    expect(reducer.visibilityFilter(oldState, action)).toEqual(newState);
  })
})