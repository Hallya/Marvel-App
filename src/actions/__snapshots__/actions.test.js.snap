// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Asyncs actions Function fetchContent:: fetch content from API and return an Object 1`] = `
Object {
  "attributionHTML": "<a href=\\"http://marvel.com\\">Data provided by Marvel. © 2018 MARVEL</a>",
  "attributionText": "Data provided by Marvel. © 2018 MARVEL",
  "code": 200,
  "copyright": "© 2018 MARVEL",
  "data": Object {
    "count": 1,
    "limit": 20,
    "offset": 0,
    "results": Array [
      Object {
        "characters": Object {
          "available": 39,
          "collectionURI": "http://gateway.marvel.com/v1/public/comics/2539/characters",
          "items": Array [
            Object {
              "name": "Abyss",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009149",
            },
            Object {
              "name": "Abyss (Age of Apocalypse)",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1010903",
            },
            Object {
              "name": "Apocalypse",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009156",
            },
            Object {
              "name": "Archangel",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009159",
            },
            Object {
              "name": "Aurora",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009163",
            },
            Object {
              "name": "Beast",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009175",
            },
            Object {
              "name": "Bishop",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009182",
            },
            Object {
              "name": "Blink",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009197",
            },
            Object {
              "name": "Blob",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009199",
            },
            Object {
              "name": "Captain Britain",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009223",
            },
            Object {
              "name": "Colossus",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009243",
            },
            Object {
              "name": "Cyclops",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009257",
            },
            Object {
              "name": "Emma Frost",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009310",
            },
            Object {
              "name": "Gambit",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009313",
            },
            Object {
              "name": "Havok",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009337",
            },
            Object {
              "name": "Holocaust (Age of Apocalypse)",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009349",
            },
            Object {
              "name": "Iceman",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009362",
            },
            Object {
              "name": "Jean Grey",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009327",
            },
            Object {
              "name": "Magma (Amara Aquilla)",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009416",
            },
            Object {
              "name": "Magneto",
              "resourceURI": "http://gateway.marvel.com/v1/public/characters/1009417",
            },
          ],
          "returned": 20,
        },
        "collectedIssues": Array [
          Object {
            "name": "Amazing X-Men (1995) #2",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/17732",
          },
          Object {
            "name": "Factor X (1995) #2",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/20045",
          },
          Object {
            "name": "Age of Apocalypse: The Chosen (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/17701",
          },
          Object {
            "name": "Weapon X (1995) #2",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/18100",
          },
          Object {
            "name": "Gambit and the X-Ternals (1995) #2",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/16078",
          },
          Object {
            "name": "Gambit and the X-Ternals (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/16077",
          },
          Object {
            "name": "Weapon X (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/18099",
          },
          Object {
            "name": "Generation Next (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/20038",
          },
          Object {
            "name": "Amazing X-Men (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/17731",
          },
          Object {
            "name": "X-Calibre (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/18253",
          },
          Object {
            "name": "Factor X (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/20044",
          },
          Object {
            "name": "X-Men: Alpha (1995) #1",
            "resourceURI": "http://gateway.marvel.com/v1/public/comics/12386",
          },
        ],
        "collections": Array [],
        "creators": Object {
          "available": 53,
          "collectionURI": "http://gateway.marvel.com/v1/public/comics/2539/creators",
          "items": Array [
            Object {
              "name": "Comicraft",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/807",
              "role": "letterer",
            },
            Object {
              "name": "Pat Brosseau",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/3626",
              "role": "letterer",
            },
            Object {
              "name": "Chris Eliopoulos",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/452",
              "role": "letterer",
            },
            Object {
              "name": "Toinay",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/5991",
              "role": "inker",
            },
            Object {
              "name": "Terry Austin",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/1836",
              "role": "inker",
            },
            Object {
              "name": "Mark Buckingham",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/71",
              "role": "inker",
            },
            Object {
              "name": "Kevin Conrad",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/972",
              "role": "inker",
            },
            Object {
              "name": "Chris Bachalo",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/232",
              "role": "penciler",
            },
            Object {
              "name": "Ian Churchill",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/1192",
              "role": "penciler",
            },
            Object {
              "name": "Tony Daniel",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/354",
              "role": "penciler",
            },
            Object {
              "name": "Steve Buccellato",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/1909",
              "role": "colorist",
            },
            Object {
              "name": "Digital Chameleon",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/2974",
              "role": "colorist",
            },
            Object {
              "name": "Electric Crayon",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/4101",
              "role": "colorist",
            },
            Object {
              "name": "Kelly Corvese",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/3572",
              "role": "editor",
            },
            Object {
              "name": "Tom DeFalco",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/9565",
              "role": "editor",
            },
            Object {
              "name": "Suzanne Gaffney",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/2099",
              "role": "editor",
            },
            Object {
              "name": "Jaye Gardner",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/2520",
              "role": "editor",
            },
            Object {
              "name": "Roger Cruz",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/356",
              "role": "penciller",
            },
            Object {
              "name": "Steve Epting",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/374",
              "role": "penciller",
            },
            Object {
              "name": "Warren Ellis",
              "resourceURI": "http://gateway.marvel.com/v1/public/creators/676",
              "role": "writer",
            },
          ],
          "returned": 20,
        },
        "dates": Array [
          Object {
            "date": "2005-10-05T00:00:00-0400",
            "type": "onsaleDate",
          },
          Object {
            "date": "-0001-11-30T00:00:00-0500",
            "type": "focDate",
          },
        ],
        "description": "See your favorite mutants through a dark glass as the epic that literally rebuilt the X-Men in eight miniseries and more continues! Apocalypse has conquered half of humankind and is ready to destroy them all! Magneto and his Amazing X-Men fight to protect humans and mutants alike, only to learn from Bishop (now on his THIRD reality and counting) that his world might need to be unmade! Plus: excerpts from Apocalypse's own files on the alternate-universe X-Men, and their friends and foes...but which is which? Featuring Blink and Sabretooth of the Exiles! Collects X-MEN: ALPHA, AGE OF APOCALYPSE: THE CHOSEN, GENERATION NEXT #1, ASTONISHING X-MEN VOL. 1 #1, X-CALIBRE #1, GAMBIT AND THE X-TERNALS #1-2, WEAPON X VOL. 1 #1-2, AMAZING X-MEN #1-2, FACTOR X #1-2 and X-MAN #1
<br>376 PGS./T+ SUGGESTED FOR TEENS AND UP ...$29.99
<br>",
        "diamondCode": "",
        "digitalId": 0,
        "ean": "",
        "events": Object {
          "available": 1,
          "collectionURI": "http://gateway.marvel.com/v1/public/comics/2539/events",
          "items": Array [
            Object {
              "name": "Age of Apocalypse",
              "resourceURI": "http://gateway.marvel.com/v1/public/events/227",
            },
          ],
          "returned": 1,
        },
        "format": "Trade Paperback",
        "id": 2539,
        "images": Array [
          Object {
            "extension": "jpg",
            "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/b0/4bc61dec7755f",
          },
        ],
        "isbn": "0-7851-1874-8",
        "issn": "",
        "issueNumber": 0,
        "modified": "2017-07-10T17:28:37-0400",
        "pageCount": 0,
        "prices": Array [
          Object {
            "price": 9.99,
            "type": "printPrice",
          },
        ],
        "resourceURI": "http://gateway.marvel.com/v1/public/comics/2539",
        "series": Object {
          "name": "X-Men: The Complete Age of Apocalypse Epic Book 2 (2005)",
          "resourceURI": "http://gateway.marvel.com/v1/public/series/1583",
        },
        "stories": Object {
          "available": 45,
          "collectionURI": "http://gateway.marvel.com/v1/public/comics/2539/stories",
          "items": Array [
            Object {
              "name": "A Beginning",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/26281",
              "type": "interiorStory",
            },
            Object {
              "name": "Some of Us Looking to the Stars",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/32819",
              "type": "interiorStory",
            },
            Object {
              "name": "X-Facts: The Dream is Over",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/32820",
              "type": "text article",
            },
            Object {
              "name": "Where No External Has Gone Before",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/32822",
              "type": "interiorStory",
            },
            Object {
              "name": "X-Facts: Reality Checked",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/32823",
              "type": "text article",
            },
            Object {
              "name": "[opening]",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37493",
              "type": "",
            },
            Object {
              "name": "Magneto and Rogue",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37494",
              "type": "",
            },
            Object {
              "name": "X-Man",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37495",
              "type": "",
            },
            Object {
              "name": "Cyclops",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37496",
              "type": "",
            },
            Object {
              "name": "Havok",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37497",
              "type": "",
            },
            Object {
              "name": "Mr. Sinister",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37498",
              "type": "",
            },
            Object {
              "name": "Storm",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37499",
              "type": "",
            },
            Object {
              "name": "Quicksilver",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37500",
              "type": "",
            },
            Object {
              "name": "Northstar and Aurora",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37501",
              "type": "",
            },
            Object {
              "name": "Beast",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37502",
              "type": "",
            },
            Object {
              "name": "The Four Horsemen",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37503",
              "type": "",
            },
            Object {
              "name": "X-Calibre",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37504",
              "type": "",
            },
            Object {
              "name": "The Human High Council",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37505",
              "type": "",
            },
            Object {
              "name": "Weapon X and Jean Grey",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37506",
              "type": "",
            },
            Object {
              "name": "Gambit and the Externals",
              "resourceURI": "http://gateway.marvel.com/v1/public/stories/37507",
              "type": "",
            },
          ],
          "returned": 20,
        },
        "textObjects": Array [
          Object {
            "language": "en-us",
            "text": "See your favorite mutants through a dark glass as the epic that literally rebuilt the X-Men in eight miniseries and more continues! Apocalypse has conquered half of humankind and is ready to destroy them all! Magneto and his Amazing X-Men fight to protect humans and mutants alike, only to learn from Bishop (now on his THIRD reality and counting) that his world might need to be unmade! Plus: excerpts from Apocalypse's own files on the alternate-universe X-Men, and their friends and foes...but which is which? Featuring Blink and Sabretooth of the Exiles! Collects X-MEN: ALPHA, AGE OF APOCALYPSE: THE CHOSEN, GENERATION NEXT #1, ASTONISHING X-MEN VOL. 1 #1, X-CALIBRE #1, GAMBIT AND THE X-TERNALS #1-2, WEAPON X VOL. 1 #1-2, AMAZING X-MEN #1-2, FACTOR X #1-2 and X-MAN #1
<br>376 PGS./T+ SUGGESTED FOR TEENS AND UP ...$29.99
<br>",
            "type": "issue_solicit_text",
          },
        ],
        "thumbnail": Object {
          "extension": "jpg",
          "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/b0/4bc61dec7755f",
        },
        "title": "X-Men: The Complete Age of Apocalypse Epic Book 2 (Trade Paperback)",
        "upc": "5960611874-00111",
        "urls": Array [
          Object {
            "type": "detail",
            "url": "http://marvel.com/comics/collection/2539/x-men_the_complete_age_of_apocalypse_epic_book_2_trade_paperback?utm_campaign=apiRef&utm_source=489e135d84bbc1a810a3a8ec86eef28e",
          },
        ],
        "variantDescription": "",
        "variants": Array [],
      },
    ],
    "total": 1,
  },
  "etag": "7692cde38ff8078a08521245b44662b9b316b1e6",
  "status": "Ok",
}
`;